osbrain/tests/test_agent.py

```diff
@@ -105,6 +105,88 @@ def test_agent_shutdown(nsproxy):
     assert 'a0' not in nsproxy.list()
 
 
+def test_close_socket_agent_shutdown(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the agent
+    is shut down
+    """
+    agent = run_agent('name')
+    address = agent.bind('PUSH')
+    agent.shutdown()
+    assert not address.address.exists()
+
+
+def test_close_socket_agent_kill(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the agent
+    is killed
+    """
+    agent = run_agent('name')
+    address = agent.bind('PUSH')
+    agent.kill()
+    assert not address.address.exists()
+
+
+def test_close_socket_agent_blocked(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed when a blocked
+    agent is requested to shut down
+    """
+    def log(agent, msg):
+        time.sleep(3)
+        return msg
+
+    alice = run_agent('alice')
+    bob = run_agent('bob')
+
+    addr = alice.bind('REP', alias='main', handler=log)
+    bob.connect(addr, alias='main')
+
+    Timer(1, nsproxy.shutdown).start()
+    bob.send('main', 'Hello world')
+    rep = bob.recv('main')
+    print(rep)
+
+    assert not addr.address.exists()
+
+
+def test_close_socket_agent_crash(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the agent
+    crashes
+    """
+    agent = run_agent('agent')
+    addr = agent.bind('PUSH', 'main')
+    with pytest.raises(RuntimeError):
+        agent.raise_exception()
+
+    assert not addr.address.exists()
+
+
+def test_close_socket_nameserver_shutdown(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the name
+    server is shut down
+    """
+    agent = run_agent('agent')
+    addr = agent.bind('PUSH', 'main')
+    nsproxy.shutdown()
+    assert not addr.address.exists()
+
+
+def test_close_socket_nameserver_crash(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the name
+    server crashes
+    """
+    agent = run_agent('agent')
+    addr = agent.bind('PUSH', 'main')
+    with pytest.raises(RuntimeError):
+        nsproxy.raise_exception()
+
+    assert not addr.address.exists()
```

Comment:
Maybe we will end-up separating this nameserver-crash test with a new issue. Not sure how we should handle this case. Maybe if the name server crashes we should shutdown everything.

Reply:
Yeah, but how are we going to do that? Maybe there's a way to modify the `run` method so we can catch exceptions and shut down the name server (or maybe try to "revive" it), but I can't think of a way to do that.

Reply:
@ocaballeror Yeah, that is why I suggested to create a separate issue. I am not sure how we could approach this either. Nameserver should *really* never crash (and has never crashed for us when using osBrain). So I do not think it is critical right now. We will need to have a look at it at some point, though. :wink:

Reply:
Definitely 👍 

I will remove the test, then. We'll add it again whenever we need to take on that issue.